examples/proof.dk

```diff
@@ -73,41 +73,92 @@ eq : !a, T a -> T a -> Prop.
 
 refl : !a, !x, P (eq a x x).
 
-eqind : !a, !x, !y, P (eq a x y) -> !p : T a -> Prop, P (p x) -> P (p y).
+eqind : !a, !x, !y, P (eq a x y) -> !p : T a -> Prop, P (p y) -> P (p x).
 (; FIXME: try to infer the type of p. ;)
 
-(; s is compatible with Leibniz equality ;)
+(; Symmetry of the equality ;)
+#PROOF eq_sym : !a, !x, !y, P (eq a x y) -> P (eq a y x).
+#REFINE a => x => y => H => _.
+(; First option, doing the proof via rewrite ;)
+#REWRITE H.
+#REFINE (refl a y).
+#QED.
+(; Second option, by hand, and full ;)
+(;
+#REFINE eqind a x y H (z => eq a y z) (refl a y).
+#QED.
+;)
+(; Third option, by hand, with a wildcard ;)
+(;
+#REFINE eqind a x y H (z => eq a y z) _.
+#SIMPL.
+#REFINE refl a y.
+#QED.
+;)
+
 
+(; s is compatible with Leibniz equality ;)
 #PROOF s_eq : !x, !y, P (eq nat x y) -> P (eq nat (s x) (s y)).
 #REFINE x => y => xy => _.
-#REFINE eqind nat x y xy (z => eq nat (s x) (s z)) (refl nat (s x)).
+#REFINE eqind nat x y xy (z => eq nat (s z) (s y)) (refl nat (s y)).
 #QED.
 
-(; 0 is right neutral for add. ;)
 
+(; 0 is right neutral for add. ;)
 #PROOF add0r : !n, P (eq nat (add n 0) n).
 (;FIXME: try to infer nat. ;)
-
 #INFER nat_ind.
 #REFINE (nat_ind (n => eq nat (add n 0) n) ?case0 ?caseS).
 (;FIXME: try to infer predicate. ;) #PRINT.
-
 #FOCUS 1. #PRINT.
 #FOCUS 1. #PRINT.
-
 (; Case 0 ;)
 #SIMPL. #PRINT.
 #REFINE (refl nat 0). #PRINT.
-
 (; Case S ;)
 #REFINE n => hn => _. #PRINT.
 #SIMPL. #PRINT.
 #REFINE s_eq (add n 0) n hn.
+#QED.
 
+
+(; forall n m, n + S m = S (n+m) ;)
+#PROOF add_succ_r : !n, !m, P (eq nat (add n (s m)) (s (add n m))).
+#REFINE n => m => _.
+#REFINE (nat_ind (n:N => eq nat (add n (s m)) (s (add n m))) ?case0[n,m] ?caseS[n,m] n).
+(; Case 0 ;)
+#SIMPL.
+#REFINE refl nat (s m). (; TO DO : ADD a #REFL tactic ;)
+(; Case S ;)
+#SIMPL.
+#REFINE pn => ih => _.
+#REWRITE ih.
+#SIMPL.
+#REFINE refl nat (s (s (add pn m))).
 #QED.
 
 
+(; Commutativity of the addition ;)
 #PROOF addcomm : !n, !m, P (eq nat (add n m) (add m n)).
 #REFINE n => m => _.
-#REFINE (nat_ind (n:N => eq nat (add n m) (add m n)) ?case0[m,n] ?caseS[m,n] n).
+#REFINE (nat_ind (n:N => eq nat (add n m) (add m n)) ?case0[n,m] ?caseS[n,m] n).
+(; Case 0 ;)
+#SIMPL.
+#REFINE eq_sym nat (add m 0) m (add0r m). (; TO DO : Add a #SYM tactic ;)
+(; Case S ;)
 #SIMPL.
+#REFINE n => ih => _.
+#REWRITE ih.
+#REWRITE add_succ_r m n0.
+#REFINE refl nat (s (add m n0)).
+#QED.
+
+
+(; At the moment, we can unfortunately prove anything in the proof mode
```

Comment:
This will not work any more when you merge, so it will probably need to be removed.

Reply:
We removed this when we added rewrtie.ml to the project. Some warnings of the sort

+ ocamlfind ocamldoc -dump src/parser.odoc -hide-warnings -charset utf-8 -package timed -package earley.str -package earley -package bindlib -package unix -pp pa_ocaml -I src src/parser.ml
File "src/parser.ml", line 15, character 0:
lexing: empty token
File "src/parser.ml", line 18, character 0:
lexing: empty token
...

appear when making the docs. If that's OK I'll add it back.

Reply:
If you are talking about the `Parser` module in `src/lambdapi.odocl` then you are right. I removed the line in master, so you can merge again.